File: backend/app/app/models/feedback.py
import datetime
from sqlalchemy import Column, ForeignKey, Integer, String, DateTime
from sqlalchemy.sql.sqltypes import JSON
from sqlalchemy.orm import relationship

from app.db.base_class import Base


class Feedback(Base):
    __tablename__ = "feedbacks"

    id = Column(Integer, primary_key=True, index=True)
    project_id = Column(Integer, ForeignKey("projects.id", ondelete="SET NULL"), index=True)
    model_id = Column(Integer, ForeignKey("project_models.id", ondelete="SET NULL"), index=True)
    model = relationship('ProjectModel')
    dataset_id = Column(Integer, ForeignKey("datasets.id", ondelete="SET NULL"), index=True)
    dataset = relationship('Dataset')
    target_feature = Column(String)
    user_id = Column(Integer, ForeignKey("user.id"), nullable=False)
    user = relationship('User')
    created_on = Column(DateTime(timezone=True), default=datetime.datetime.now)
    feedback_type = Column(String, nullable=False, index=True)
    feature_name = Column(String)
    feature_value = Column(String)
    feedback_choice = Column(String)
    feedback_message = Column(String)
    user_data = Column(JSON, nullable=False)
    original_data = Column(JSON, nullable=False)
    replies = relationship('FeedbackReply')

File: backend/java-app/src/main/java/ai/giskard/domain/Feedback.java
package ai.giskard.domain;

import ai.giskard.domain.ml.Dataset;
import ai.giskard.domain.ml.ProjectModel;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.time.LocalDateTime;

@Entity
public class Feedback {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Getter
    @Setter
    @ManyToOne
    private Project project;

    @Getter
    @Setter
    @ManyToOne
    private ProjectModel model;

    @Getter
    @Setter
    @ManyToOne
    private Dataset dataset;

    @Getter
    @Setter
    @ManyToOne
    private User user;

    private String targetFeature;

    private LocalDateTime createdOn;
    @NotNull
    private String feedbackType;
    private String featureName;
    private String featureValue;
    private String feedbackChoice;
    private String feedbackMessage;
    @NotNull
    private String userData;
    @NotNull
    private String originalData;
}
